### ISSUE TITLE
MS Word, some word spaces get lost (line break issue?)

### ISSUE THREAD

**Issue by yankustefan** [fraywing/textAngular#598]
Related to "Pasting from word #447"

When I paste some text from MS Word, some spaces (between words) get lost.

In search of the problematic character, I tried to get hold of the data in the clipboard, but I can't figure it out.

However, if I paste with ta-unsafe-sanitizer, and look at the HTML, there are line breaks, where there should be spaces.

These line breaks do not correspond with line breaks in MS Word.
(It can be the second word in a new paragraph)
See Images below.

I wonder if it would be best to replace these line brakes with spaces, and get rid of double spaces later.

Image of HTML output (with ta-unsafe-sanitizer="true"):
![ta-spaces](https://cloud.githubusercontent.com/assets/8319813/6818125/a55f093c-d2ed-11e4-965b-ad148496586a.jpg)

Image of EDITOR output (with ta-unsafe-sanitizer="true"):
![ta-spaces_editor](https://cloud.githubusercontent.com/assets/8319813/6818151/10677408-d2ee-11e4-82ea-7fed9c63948f.jpg)

Image of HTML output (with ta-unsafe-sanitizer="false"):
![ta-spaces_editor_2](https://cloud.githubusercontent.com/assets/8319813/6818210/38787c70-d2ef-11e4-99b5-21c9a5f9adde.jpg)

Visible result is the same.

**Comment by zigrivers** [comment]
I'm running into this same problem. The workaround that I'm having our users do until this is fixed is to paste the text from MS Word into another editor (e.g. Sublime) and then paste into textAngular. They lose all their formatting and have to reformat using textAngular, which is a pain, but I couldn't figure out a way around this bug short of using a different control and I don't want to do that.

**Comment by hcharania** [comment]
We are also facing this issue with our users.  Just pasted a paragraph from Word into the online Demo page, and confirmed that the problem exists there too.

**Comment by snugbeard** [comment]
I noticed that MS Word (here: 2013) "generate" Unicode line breaks as &#10; chars. Those are stripped by textAngular somehow during sanatizing I suppose.

Strangly those line breaks are not in the original document and are added while copying formatted text from Word. So, this is somehow a MS Word issue.

A workaround would be to replace &#10; with </br> but that would change the actual text layout from MS Word. But at least the User is able to see those; missing spaces are hard to track.

A clean solution is more difficult. textAngular would need to recognize MS Word generated content e.g. replacing &#10; with a space.

Or maybe you can think of a better solution...

**Comment by yankustefan** [comment]
@snugbeard, thanks for investigating.

**Comment by janvandenberg** [comment]
I have the same issue with copy+paste from MS Word. @snugbeard 's solution works for me too, but I would love to see a permanent solution in an upcoming version of textAngular.

**Comment by jeserkin** [comment]
+1

**Comment by 0AlexScott** [comment]
+1 for @snugbeard 's workaround - anyone came across any side effects from this change?

**Comment by davetheflashguy** [comment]
+1 as well for @snugbeard 's workaround.  The only reason I could see this solution as being a problem is if you intend to display the text in a pre tag.  While I only use textAngular for a small part of my application and I also only use a few of the toolbar options, I have done a full regression test using existing entries into the editor and have not found any problems at all after implementing this fix.

Thanks for the tip!

**Comment by JoelParke** [comment]
<p>Could you retest this with release 1.4.6 which I will put out in a few minutes... Thanks, Joel</p>

**Comment by jeserkin** [comment]
@JoelParke will it be updated on [official webpage](http://textangular.com/) as well?

**Comment by JoelParke** [comment]
<p>Everything is up to 1.4.6 at this point.  Or if it is not, please let me know where it's broken.  npm and bower are also updated. <br/>Please retest and let me know....</p><p>Thanks, <br/>Joel</p>

**Comment by stevokk** [comment]
I was seeing this issue a while ago too, Word was using the LF character instead of a space sometimes. I am unable to test @snugbeard's fix as I no longer have a document with such formatting but this was with 1.4.6 so if someone in this thread has such a file, can they run it through to see if the bug is resolved for them?

My fix was to add code after line 1563:
```
text = targetDom.html();
// LF characters instead of spaces in some spots and they are replaced by "/n", so we need to just swap them to spaces 
text = text.replace(/\n/g, ' ')
```

**Comment by JoelParke** [comment]
<p>This fix seems to work very well.  I will be testing and incorporate into into the next release!</p><p>Thanks very much for pointing this out.   Sorry that I took so long to get to. </p>

**Comment by JoelParke** [comment]
<p>This will be fixed in release 1.5.12</p>
